File: 1110_Delete Nodes And Return Forest/DeleteNodesAndReturnForest.java
import java.util.*;

public class Solution {
    public List<TreeNode> delNodes(TreeNode root, int[] to_delete) {
        List<TreeNode> ans = new LinkedList<TreeNode>();
        Set<Integer> set = new HashSet<Integer>();
        for (int v : to_delete) {
            set.add(v);
        }
        postOrder(root, set, ans);
        if (!set.contains(root.val))
            ans.add(root);
        return ans;
    }

    private TreeNode postOrder(TreeNode node, Set set, List<TreeNode> ans) {
        if (node == null)
            return null;
        node.left = postOrder(node.left, set, ans);
        node.right = postOrder(node.right, set, ans);
        if (!set.contains(node.val))
            return node;
        if (node.left != null)
            ans.add(node.left);
        if (node.right != null)
            ans.add(node.right);
        return null;
    }
}

File: 1110_Delete Nodes And Return Forest/DeleteNodesAndReturnForest.cs
public class Solution {
    public IList<TreeNode> DelNodes(TreeNode root, int[] to_delete) {
        var ans = new List<TreeNode>();
        var toDeleteSet = new HashSet<int>(to_delete);
        var node = Traverse(root, ans, toDeleteSet);
        if(node != null) ans.Add(node);
        return ans;
    }
    
    private TreeNode Traverse(TreeNode node, IList<TreeNode> ans, HashSet<int> toDeleteSet)
    {
        if(node == null) return null;
        node.left = Traverse(node.left, ans, toDeleteSet);
        node.right = Traverse(node.right, ans, toDeleteSet);
        
        if(toDeleteSet.Contains(node.val))
        {
            if(node.left != null) ans.Add(node.left);
            if(node.right != null) ans.Add(node.right);
            return null;
        }else
            return node;
    }
}